feat: simplify channel accounting, refactor state transitions

kincaidoneil opened the pull request:
#### Balances

- Balance logic is a no-op on client, must call `sendMoney`
- No balance config required on server: by default, requires prefunding & auto settles to 0
- Different client/server behavior temporarily complicates the balance logic until the connector refactor is complete
- Channels are not auto created on client and channel fees are not accounted in the balance (much simpler accounting, but not a permanent solution)
- BREAKING: balances for existing accounts are incompatible

#### Internal refactor

- Dual queues for incoming/outgoing channels
- All channel/claim _state_ updates are queued (previously a shared cached would be mutated, but these updates weren't queued and were hard to reason about)
- Single data structure for channels & claims
- State transitions much more straightforward (but will continue to iterate)
- BREAKING: new channel/claim schema
- BREAKING: claims are sent as BTP type MESSAGE instead of type TRANSFER

(Will add the auto-funding functionality to the server at a later commit)

kincaidoneil committed:
chore: add prettier for formatting

kincaidoneil committed:
feat: no auto channel funding, refactor channel state transitions

Balances:
- Balance logic is a no-op on client, must call sendMoney
- No balance config required on server: by default, requires prefunding & auto settles to 0
- Different client/server behavior temporarily complicates the balance logic until the connector refactor is complete
- Channels are not auto created on client and fees are not accounted in the balance (much simpler accounting!)
- BREAKING: balances for existing accounts are incompatible

Internal refactor:
- Dual queues for incoming/outgoing channels
- All channel/claim state updates are queued
- Single data structure for channels & claims
- State transitions much easier to reason about, but will continue to iterate
- BREAKING: new channel/claim schema
- BREAKING: claims are sent as BTP MESSAGE instead of TRANSFER

kincaidoneil committed:
test: update, simplify, remove stream

kincaidoneil committed:
fix: skipped test, lint error, missing dep

kincaidoneil committed:
fix: requestClose, export some utils

- Requests to peer to close a channel will refresh the state until it's closed
- Fixs bugs claiming channel after receiving request close message from peer
- enable multiple commonjs exports

kincaidoneil committed:
feat: auto fund outgoing capacity

- adds minIncomingChannelAmount config option
- by default, no channels are automatically funded (minIncoming=Infinity)
- if an incoming channel value is > minIncoming, an outgoing channel will be funded
- if enabled, the outgoing channel will automatically be topped up when it gets half empty
- minor bug fixes
- after opening channel, sends 0 amount claim as "proof of channel" so the peer links it

kincaidoneil committed:
3.0.0-beta.12

karzak commented on src/account.ts:
```diff
@@ -1,645 +1,733 @@
+
+  /**
+   * Amount of failed outgoing settlements that is owed to the peer, but not reflected
+   * in the payableBalance (e.g. due to sendMoney calls on client)
```
I find this description confusing, don't know how to make it better though.

kincaidoneil replied:
Yeah :/ Hopefully it'll become simpler after the payable/receivable balances move to the connector.

karzak commented on src/index.ts:
```diff
@@ -305,19 +321,27 @@ export default class EthereumPlugin extends EventEmitter2 implements PluginInsta
+  async sendMoney(amount: string) {
+    const peerAccount = this._accounts.get('peer')
+    if (peerAccount) {
+      // If the plugin is acting as a client, enable sendMoney (required for prefunding)
```
Should update the README, which currently says sendMoney is always a no-op.

karzak commented on src/account.ts:
```diff
@@ -1,645 +1,733 @@
+        return cachedChannel
+          ? this.depositToChannel(
+              cachedChannel,
+              this.master._outgoingChannelAmount.minus(cachedChannel.value)
```
```suggestion
              this.master._outgoingChannelAmount.minus(remainingInChannel(cachedChannel))
```

kincaidoneil replied:
Resolved by 39c1c48e31a33497d73c20909e6a0ff0ce7c6e3e

kincaidoneil committed:
fix: top up channel for correct amount

- bug would trigger deposits of 0

kincaidoneil committed:
BREAKING: only support/test node > 10

kincaidoneil committed:
feat: tx submitter, parallel deposits and claims

- abstract channel refreshing and tx submission code
- use queue shared by all accounts to submit txs one at a time
- create separate queue for outgoing claims used *only* while a deposit is in progress, enabling them to happen in parallel

kincaidoneil committed:
3.0.0-beta.13

kincaidoneil committed:
fix: bug deserializing stored paychans, await contract metadata

kincaidoneil committed:
3.0.0-beta.14

kincaidoneil committed:
docs(readme): update with config changes

kincaidoneil committed:
fix: less logs when retrying claim validation
